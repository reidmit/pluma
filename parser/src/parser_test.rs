use crate::{Parser, Tokenizer};

/// These tests cover the basic functionality of the parser.
///
/// See the snapshots/ directory in this crate for the snapshot output. The files tend
/// to be named like `{crate}__{file}__{test_name}.snap` - for example, see
/// `parser__parser_test__hello_world.snap`.

macro_rules! snapshot {
	($test_name:ident, $raw_source:literal) => {
		#[test]
		fn $test_name() {
			let source = $raw_source;
			let bytes = source.as_bytes().to_vec();
			let tokenizer = Tokenizer::from_source(&bytes);
			let mut parser = Parser::new(&bytes, tokenizer);
			let result = parser.parse_module();

			insta::assert_snapshot!(format!("{}\n\n{:#?}", source, result));
		}
	};
}

snapshot!(hello_world, "hello world");
