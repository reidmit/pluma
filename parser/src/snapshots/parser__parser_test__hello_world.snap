---
source: parser/src/parser_test.rs
expression: "format!(\"{}\\n\\n{:#?}\", source, result)"
---
hello world

(
    ModuleNode {
        pos: (
            0,
            11,
        ),
        body: [
            StatementNode {
                pos: (
                    0,
                    11,
                ),
                kind: Expr(
                    ExprNode {
                        pos: (
                            0,
                            11,
                        ),
                        kind: Call {
                            call: CallNode {
                                pos: (
                                    0,
                                    11,
                                ),
                                callee: ExprNode {
                                    pos: (
                                        0,
                                        5,
                                    ),
                                    kind: Identifier {
                                        ident: IdentifierNode {
                                            pos: (
                                                0,
                                                5,
                                            ),
                                            name: "hello",
                                        },
                                    },
                                    typ: Unknown,
                                },
                                args: [
                                    ExprNode {
                                        pos: (
                                            6,
                                            11,
                                        ),
                                        kind: Identifier {
                                            ident: IdentifierNode {
                                                pos: (
                                                    6,
                                                    11,
                                                ),
                                                name: "world",
                                            },
                                        },
                                        typ: Unknown,
                                    },
                                ],
                                typ: Unknown,
                            },
                        },
                        typ: Unknown,
                    },
                ),
            },
        ],
    },
    (
        {},
        [],
    ),
    [],
)
